File: balancin/python/sim_mpc.py
"""
#2  MPC / LQR PREDICTIVO para el balancin.
Convencion del curso (control_procesos/MPC_funcion): c2d + horizonte N + costo cuadratico.
Version verificable: LQR de horizonte finito por recursion de Riccati hacia atras
(= "LQR predictivo"); la ganancia del 1er paso K_mpc es la que se implementa.
El QP con restricciones (|u|<=255) se resuelve por saturacion (ver nota MATLAB).
"""
import numpy as np
from scipy.linalg import solve_discrete_are
import matplotlib; matplotlib.use("Agg"); import matplotlib.pyplot as plt
import os,sys; sys.path.insert(0,os.path.dirname(__file__)); from planta import plant

A,B,C,D = plant(); n=4; R2D=180/np.pi; dt=0.01
# --- discretizar (ZOH aprox) ---
Ad = np.eye(n)+A*dt+0.5*(A@A)*dt**2
Bd = (np.eye(n)*dt+0.5*A*dt**2)@B
# --- pesos MPC ---
Q  = np.diag([1.,1.,300.,10.]);  Rw = np.array([[0.0002]]);  Np = 60
# --- Riccati hacia atras en horizonte Np (LQR predictivo) ---
P = Q.copy()
for _ in range(Np):
    K = np.linalg.inv(Rw + Bd.T@P@Bd) @ (Bd.T@P@Ad)      # ganancia paso a paso
    P = Q + Ad.T@P@(Ad - Bd@K)
Kmpc = K.flatten()                                        # u = -Kmpc x  (1er paso)
print("Kmpc =", np.round(Kmpc,3))
print(f"  Kang_p={Kmpc[2]/R2D:.2f} PWM/deg  Kang_d={Kmpc[3]/R2D:.2f}  Kpos_p={Kmpc[0]:.2f}  Kpos_d={Kmpc[1]:.2f}")

# --- simulacion lazo cerrado con saturacion (restriccion de entrada) ---
N=int(3/dt); x=np.array([0,0,5/R2D,0.]); X=np.zeros((N,4)); U=np.zeros(N)
for k in range(N):
    u=np.clip(float(-Kmpc@x),-255,255); X[k]=x; U[k]=u
    x=Ad@x+Bd.flatten()*u
t=np.arange(N)*dt
print("ESTABLE:", abs(X[-1,2]*R2D)<1 and np.max(np.abs(X[:,2]))<0.6,
      " theta_f=%.3f deg |u|max=%.0f"%(X[-1,2]*R2D,np.max(np.abs(U))))
fig,ax=plt.subplots(2,1,figsize=(9,5),sharex=True)
ax[0].plot(t,X[:,2]*R2D); ax[0].set_ylabel("theta[deg]"); ax[0].grid(alpha=.3); ax[0].set_title("MPC / LQR predictivo")
ax[1].plot(t,U); ax[1].set_ylabel("u[PWM]"); ax[1].set_xlabel("t[s]"); ax[1].grid(alpha=.3)
plt.tight_layout(); plt.savefig(os.path.join(os.path.dirname(__file__),"..","data","sim_mpc.png"),dpi=100)

File: balancin/matlab/sim_mpc.m
clc; clear; close all;               % #2 MPC / LQR PREDICTIVO
[A,B,C,D]=planta(); R2D=180/pi; dt=0.01;
sysd=c2d(ss(A,B,C,D),dt,'zoh'); Ad=sysd.A; Bd=sysd.B;
Q=diag([1 1 300 10]); Rw=0.0002; Np=60;
% --- LQR de horizonte finito (Riccati hacia atras) ---
P=Q;
for i=1:Np
  K=(Rw+Bd'*P*Bd)\(Bd'*P*Ad); P=Q+Ad'*P*(Ad-Bd*K);
end
Kmpc=K; fprintf('Kmpc=[%.2f %.2f %.2f %.2f]\n',Kmpc);
% --- (opcional) QP con restricciones estilo tu MPC_funcion:
%   quadprog con Umax/Umin. Aqui usamos saturacion de la ganancia. ---
x=[0;0;5/R2D;0]; N=round(3/dt); X=zeros(4,N); U=zeros(1,N);
for k=1:N
  u=max(min(-Kmpc*x,255),-255); X(:,k)=x; U(k)=u; x=Ad*x+Bd*u;
end
t=(0:N-1)*dt;
subplot(2,1,1);plot(t,X(3,:)*R2D);ylabel('\theta[deg]');grid on;title('MPC/LQR predictivo');
subplot(2,1,2);plot(t,U);ylabel('u[PWM]');xlabel('t[s]');grid on;
